storages/memory-storage/src/transaction.rs

```diff
@@ -10,20 +10,41 @@ use {
 #[async_trait(?Send)]
 impl Transaction for MemoryStorage {
     async fn begin(&mut self, autocommit: bool) -> Result<bool> {
-        if autocommit {
-            return Ok(false);
-        }
+        let current_state = self.state.clone();
+        match (current_state, autocommit) {
+            //TODO : implement nested transaction later
+            (StorageState::Transaction { .. }, false) => Err(Error::StorageMsg(
+                "nested transaction is not supported".to_owned(),
+            )),
+            (StorageState::Transaction { autocommit }, true) => Ok(autocommit),
+            (StorageState::Idle, _) => {
+                self.set_snapshot();
+                self.state = StorageState::Transaction { autocommit: false };
 
-        Err(Error::StorageMsg(
-            "[MemoryStorage] transaction is not supported".to_owned(),
-        ))
+                Ok(autocommit)
+            }
+        }
     }
 
     async fn rollback(&mut self) -> Result<()> {
+        match &self.snapshot {
+            Some(snapshot) => {
+                let data = snapshot.to_owned().data;
+                self.functions = data.functions;
+                self.id_counter = data.id_counter;
+                self.items = data.items;
+                self.metadata = data.metadata;
+                self.state = StorageState::Idle;
+                self.snapshot = None;
+            }
+            None => println!("TODO: non snapshot error handling"),
```

Comment:
Are you on way to implement this part?
And you can remove other `TODO` comments if was done.

Reply:
Apologize for incomplete pull request. 
I'm working on optimizing transaction functionality now and will update this pull request based on your review soon!